paasta_tools/list_marathon_service_instances.py

```diff
@@ -99,27 +104,31 @@ def get_desired_marathon_configs(soa_dir):
             )
         except (NoDeploymentsAvailable, NoDockerImageError):
             pass
-    return marathon_configs
+    return formatted_marathon_configs, job_configs
 
 
 @use_requests_cache('list_marathon_services')
-def get_service_instances_that_need_bouncing(marathon_client, soa_dir):
-    desired_marathon_configs = get_desired_marathon_configs(soa_dir)
-    desired_ids = set(desired_marathon_configs.keys())
+def get_service_instances_that_need_bouncing(marathon_clients, soa_dir):
+    desired_marathon_configs_formatted, desired_job_configs = get_desired_marathon_configs(soa_dir)
+    desired_ids_and_clients = set()
+    for app_id, job_config in desired_job_configs.items():
+        desired_ids_and_clients.add((app_id, marathon_clients.get_current_client_for_service(job_config)))
 
-    current_apps = {app.id.lstrip('/'): app for app in marathon_client.list_apps()}
-    actual_ids = set(current_apps.keys())
+    current_apps_with_clients = {}
+    for client in marathon_clients.get_all_clients():
+        current_apps_with_clients.update({(app.id.lstrip('/'), client): app for app in client.list_apps()})
+    actual_ids_and_clients = set(current_apps_with_clients.keys())
 
-    apps_that_need_bouncing = actual_ids.symmetric_difference(desired_ids)
-    apps_that_need_bouncing = {long_job_id_to_short_job_id(app_id) for app_id in apps_that_need_bouncing}
+    undesired_apps_and_clients = actual_ids_and_clients.symmetric_difference(desired_ids_and_clients)
+    apps_that_need_bouncing = {long_job_id_to_short_job_id(app_id) for app_id, client in undesired_apps_and_clients}
 
     draining_hosts = get_draining_hosts()
 
-    for app_id, app in current_apps.items():
+    for (app_id, client), app in current_apps_with_clients.items():
```

Comment:
I think there may be a subtle bug here if we get to the place where we have an appid on a old client, but we'll never get rid of it if the *actual* bounce code can't "find" it because it isn't in a previous client list.

I hope this doesn't happen, but if it does it will mean that the app is *constantly* being asked to bounce, but it will never converge. right?

Reply:
I'm not sure I follow. You mean like the scenario tested in [this itest](https://github.com/Yelp/paasta/pull/1552/files/fc9641baaf795652f35a46fde3764f418c9609f7#diff-b66d150a775e1d9eb08819b944abeec5R72)?